src/test/run-pass/generator/auxiliary/xcrate.rs

```diff
@@ -8,19 +8,20 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-#![feature(generators, generator_trait)]
+#![feature(generators, generator_trait, pin)]
 
+use std::marker::Unpin;
 use std::ops::Generator;
 
-pub fn foo() -> impl Generator<Yield = (), Return = ()> {
+pub fn foo() -> impl Generator<Yield = (), Return = ()> + Unpin {
     || {
         if false {
             yield;
         }
     }
 }
 
-pub fn bar<T: 'static>(t: T) -> Box<Generator<Yield = T, Return = ()>> {
+pub fn bar<T: Unpin + 'static>(t: T) -> Box<Generator<Yield = T, Return = ()> + Unpin> {
```

Comment:
The `Unpin` bound on `T` here is strange too.

Reply:
I think that's necessary because non-`static` generators are still subject to the normal structural `Unpin` checks. Rebuilding without it now to check (will be a while as I've got to do a fresh build).

Reply:
So, I haven't spent time verifying this, but having non-`static` generators use the structural definition of pinning makes this sound code:

```rust
fn foo() -> impl Generator<Yield = usize, Return = ()> {
    static || {
        let x = 5;
        let x = &x;
        yield *x;
        yield *x;
    }
}

fn bar() -> impl Generator<Yield = usize, Return = ()> {
    || {
        let mut foo = foo();
        loop {
            let val = {
                let mut foo = unsafe { Pin::new_unchecked(&mut foo) };
                match foo.resume() {
                    GeneratorState::Yielded(val) => val,
                    GeneratorState::Complete(()) => break,
                }
            };
            yield val;
        }
    }
}
```

I don't know if there would be a way to make this _safe_ code, the stack pinning macro would require borrows to persist across yields, but maybe something like this would be useful?

It also seems to me that this should be investigated for both closures and generators at the same time, they likely have similar reasons to either propagate or force `Unpin`.

@withoutboats any opinion (and does closures interaction with `Unpin` matter for stabilization of `Pin`)?

Reply:
There should be some way to make generators which are checked for `Unpin`, so they can be moved in and out of pins (and thus moved between resume calls). Currently its not annotating the generator `static`.

I think the current closure behavior around `Unpin` is correct in that it is consistent with other auto traits like `Send` etc; if anything should change (not sure that it should) it would be generator syntax.

Reply:
> maybe it only should check that they don't themselves borrow across yields, and omit the generated negative impl of unpin.

That's basically what is happening now, for `static` annotated generators there's an implicit negative impl of `Unpin`, for other generators there's no implicit impl (either positive or negative) and it's falling back to structural checking of the upvars and state that is live across yield points (the closed over _environment_).

We could force a positive impl of `Unpin` for non-`static` generators instead and say that it's unsound to create a pinned reference into the environment, even just between yield points. You would still be able to create pinned references to values that are only live between yield points since they are true stack variables. And if you need pinned references into the environment you can force the generator to be immovable via `static`. As far as I'm aware this wouldn't remove any capabilities; a non-`static` generator that is `!Unpin` because of structural recursion and a `static` generator appear the same externally, and when implementing the generator you always have the choice to add or remove the `static` annotation

---

Thinking about it a bit more I guess this doesn't matter for closure's as we don't have `FnPin`, there's no way to call a closure such that it could soundly create a pinned reference into its environment.

Reply:
I've added an unconditional implementation of `Unpin` for movable generators.